keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,734 @@
+This KEP does not attempt to implement that. These stats are largely considered [a part of Kubernetes](#keep-summary). The stated goal of this KEP is to reduce the overhead
+on two entities reporting metrics, not totally changing what stats the Kubelet reports (with the exception of Accelerator and UserDefinedMetrics).
+
+Thus, this KEP largely the plan described [here](#plan), with some changes:
+
+- the `/metrics/cadvisor` endpoint will likely not be exposed in the Alpha release, and its continued support will need to be evaluated for the future.
+- CRI is used for all of the monitoring endpoints related to Containers and Pods (except Volume and Ephemeral Storage)
+- CRI is used to provide metrics for eviction.
+
+[core-proposal]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/core-metrics-pipeline.md#proposed-core-metrics
+[keep-summary]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-666636130
+[plan]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-724928827
+
+#### cAdvisor Metrics Endpoint
+
+In addition to the `/stats/summary` endpoint, there are also consumers that rely on `/metrics/cadvisor`.
+In the past, each of these were populated primarily by cAdvisor, but the goals of this KEP will prevent cAdvisor from hosting these metrics.
+Where the majority of internal Kubernetes entities rely solely on the Summary API, there are likely consumers in the community that rely on this endppoint.
+
+Thus, in addition to fulfilling `/stats/summary` endpoint, this KEP will have to decide on how to handle `/metrics/cadvisor`
+Below is the proposed strategy for doing so:
+
+1. For the Alpha release of this KEP, accept there will be breaking changes to `/metrics/cadvisor`. Use a feature gate and the strategies proposed above to minimize breakages.
+2. For the Beta release, evaluate the future of the `/metrics/cadvisor` endpoint. We can either:
+    - Deprecate the endpoint, and extend the Summary API to report the required fields.
+    - Move the reporter of `/metrics/cadvisor` from cAdvisor to the CRI implementation.
+		- Note: this means CRI implementations will have to support all the metrics of this endpoint in a satisfactory manner.
+		- This does not include extensions the CRI API itself, because it will be satisfied internally by the CRI implementation.
+		- Entities relying on this endpoint would then query the CRI implementation directly, rather than go through the Kubelet.
+    - Extend the CRI and have the CRI implementation report the required stats, and have the Kubelet implement the `/metrics/cadvisor` endpoint. 
+3. For the GA release, the CRI implementation should be the source of truth for all pod and container level metrics that external parties rely on (no matter how many endpoints the Kubelet advertises).
+
+#### Open Questions
+1. For the newly introduced CRI API fields, should there be Windows and Linux specific fields?
```

Comment:
it looks like it's mostly network related stats (basing it off of cri_stats_provider_windows.go). @marosset or others, I'd like some help here, I don't have access to windows nodes to see what's reported

Reply:
Yes, I think we should put linux specific stats into a linux specific section, so that it is clear which stats are specific to linux/windows.

I think that should be incorporated in this proposal.

Reply:
Looks like we still don't have the answer for this.  Since @marosset is out, @jsturtevant, can you shed some lights here?

Reply:
@dchen1107 @kevpar If it helps this is the output of the /stats/summary on my containerd node
``` json
{
  "node": {
    "nodeName": "peri-win-md-0-windows-containerd-658ddf8989-58ng7",
    "systemContainers": [
      {
        "name": "pods",
        "startTime": "2021-05-14T23:21:23Z",
        "cpu": {
          "time": "2021-05-14T23:21:23Z",
          "usageNanoCores": 930724,
          "usageCoreNanoSeconds": 11250000000
        },
        "memory": {
          "time": "2021-05-14T23:21:23Z",
          "workingSetBytes": 54337536
        }
      }
    ],
    "startTime": "2021-05-07T21:27:47Z",
    "cpu": {
      "time": "2021-05-14T23:21:02Z",
      "usageNanoCores": 1600000000,
      "usageCoreNanoSeconds": 87813080000000
    },
    "memory": {
      "time": "2021-05-14T23:21:02Z",
      "availableBytes": 7171022848,
      "usageBytes": 3469168640,
      "workingSetBytes": 1418375168,
      "rssBytes": 0,
      "pageFaults": 0,
      "majorPageFaults": 0
    },
    "network": {
      "time": "2021-05-14T23:21:02Z",
      "name": "",
      "interfaces": [
        {
          "name": "Hyper-V Virtual Switch Extension Adapter",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Microsoft IP-HTTPS Platform Interface",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Ethernet Adapter _2",
          "rxBytes": 7705454,
          "rxErrors": 0,
          "txBytes": 996633,
          "txErrors": 0
        },
        {
          "name": "Teredo Tunneling Pseudo-Interface",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "vmxnet3 Ethernet Adapter",
          "rxBytes": 1079972033714,
          "rxErrors": 0,
          "txBytes": 194073282,
          "txErrors": 0
        },
        {
          "name": "6to4 Adapter",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Ethernet Adapter",
          "rxBytes": 996648,
          "rxErrors": 0,
          "txBytes": 7705408,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Switch Extension Adapter _2",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Ethernet Adapter _3",
          "rxBytes": 966122969568,
          "rxErrors": 0,
          "txBytes": 194040173,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Ethernet Adapter _4",
          "rxBytes": 173485,
          "rxErrors": 0,
          "txBytes": 639848,
          "txErrors": 0
        },
        {
          "name": "Microsoft Kernel Debug Network Adapter",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        }
      ]
    },
    "fs": {
      "time": "2021-05-14T23:21:02Z",
      "availableBytes": 62825033728,
      "capacityBytes": 85530243072,
      "usedBytes": 22705209344
    },
    "runtime": {
      "imageFs": {
        "time": "2021-05-14T23:20:59Z",
        "availableBytes": 62825029632,
        "capacityBytes": 85530243072,
        "usedBytes": 14650776773,
        "inodesUsed": 0
      }
    }
  },
  "pods": [
    {
      "podRef": {
        "name": "win-webserver-79447c5fcf-kktpn",
        "namespace": "default",
        "uid": "121343f4-aa2c-4bab-96fb-36d176133955"
      },
      "startTime": "2021-05-14T23:11:07Z",
      "containers": [
        {
          "name": "windowswebserver",
          "startTime": "2021-05-14T23:18:48Z",
          "cpu": {
            "time": "2021-05-14T23:21:03Z",
            "usageNanoCores": 930724,
            "usageCoreNanoSeconds": 11250000000
          },
          "memory": {
            "time": "2021-05-14T23:21:03Z",
            "workingSetBytes": 54337536
          },
          "rootfs": {
            "time": "2021-05-14T23:21:09Z",
            "availableBytes": 62635749376,
            "capacityBytes": 85530243072,
            "usedBytes": 71303168,
            "inodesUsed": 0
          },
          "logs": {
            "time": "2021-05-14T23:21:23Z",
            "availableBytes": 62825029632,
            "capacityBytes": 85530243072,
            "usedBytes": 0,
            "inodesUsed": 0
          }
        }
      ],
      "cpu": {
        "time": "2021-05-14T23:21:03Z",
        "usageNanoCores": 930724,
        "usageCoreNanoSeconds": 11250000000
      },
      "memory": {
        "time": "2021-05-14T23:21:03Z",
        "availableBytes": 0,
        "usageBytes": 0,
        "workingSetBytes": 54337536,
        "rssBytes": 0,
        "pageFaults": 0,
        "majorPageFaults": 0
      },
      "volume": [
        {
          "time": "2021-05-14T23:11:29Z",
          "availableBytes": 71143952384,
          "capacityBytes": 85530243072,
          "usedBytes": 6032,
          "inodesFree": 0,
          "inodes": 0,
          "inodesUsed": 0,
          "name": "default-token-cx6sp"
        }
      ],
      "ephemeral-storage": {
        "time": "2021-05-14T23:21:23Z",
        "availableBytes": 62825029632,
        "capacityBytes": 85530243072,
        "usedBytes": 71303168,
        "inodesUsed": 0
      }
    }
  ]
}
```